File: permissionsex-core/src/test/java/ca/stellardrift/permissionsex/util/command/GenericArgumentsTest.java
/*
 * PermissionsEx
 * Copyright (C) zml and PermissionsEx contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ca.stellardrift.permissionsex.util.command;

import ca.stellardrift.permissionsex.commands.commander.Commander;
import ca.stellardrift.permissionsex.util.Translations;
import ca.stellardrift.permissionsex.util.command.args.*;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;

import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests for all argument types contained in GenericArguments
 */
public class GenericArgumentsTest {
    static final CommandExecutor NULL_EXECUTOR = new CommandExecutor() {
        @Override
        public <TextType> void execute(Commander<TextType> src, CommandContext args) throws CommandException {
        }
    };

    private static CommandContext parseForInput(String input, CommandElement element) throws ArgumentParseException {
        CommandSpec spec = CommandSpec.builder()
                .setAliases("test")
                .setExecutor(NULL_EXECUTOR)
                .build();
        CommandArgs args = QuotedStringParser.parseFrom(input, false);
        CommandContext context = new CommandContext(spec, args.getRaw());
        element.parse(args, context);
        return context;
    }

    @Test
    public void testNone() throws ArgumentParseException {
        CommandArgs args = QuotedStringParser.parseFrom("a", false);
        CommandContext context = new CommandContext(CommandSpec.builder().setAliases("test").setExecutor(NULL_EXECUTOR).build(), args.getRaw());
        GenericArguments.none().parse(args, context);
        assertEquals("a", args.next());
    }

    @Test
    @Disabled
    public void testFlags() {
        throw new UnsupportedOperationException();
    }

    @Test
    public void testSequence() throws ArgumentParseException {
        CommandElement el = GenericArguments.seq(GenericArguments.string(Translations.untr("one")), GenericArguments.string(Translations.untr("two")), GenericArguments.string(Translations.untr("three")));
        CommandContext context = parseForInput("a b c", el);
        assertEquals("a", context.getOne("one"));
        assertEquals("b", context.getOne("two"));
        assertEquals("c", context.getOne("three"));

        Assertions.assertThrows(ArgumentParseException.class, () -> {
            parseForInput("a b", el);
        });
    }

    @Test
    public void testChoices() throws ArgumentParseException {
        CommandElement el = GenericArguments.choices(Translations.untr("val"), ImmutableMap.of("a", "one", "b", "two"));
        CommandContext context = parseForInput("a", el);
        assertEquals("one", context.getOne("val"));

        Assertions.assertThrows(ArgumentParseException.class, () -> {
            parseForInput("c", el);
        });
    }

    @Test
    public void testFirstParsing() throws ArgumentParseException {
        CommandElement el = GenericArguments.firstParsing(GenericArguments.integer(Translations.untr("val")), GenericArguments.string(Translations.untr("val")));
        CommandContext context = parseForInput("word", el);
        assertEquals("word", context.getOne("val"));

        context = parseForInput("42", el);
        assertEquals((Integer) 42, context.getOne("val"));
    }

    @Test
    public void testOptional() throws ArgumentParseException {
        CommandElement el = GenericArguments.optional(GenericArguments.string(Translations.untr("val")));
        CommandContext context = parseForInput("", el);
        assertNull(context.getOne("val"));

        el = GenericArguments.optional(GenericArguments.string(Translations.untr("val")), "def");
        context = parseForInput("", el);
        assertEquals("def", context.getOne("val"));

        el = GenericArguments.seq(GenericArguments.optionalWeak(GenericArguments.integer(Translations.untr("val"))), GenericArguments.string(Translations.untr("str")));
        context = parseForInput("hello", el);
        assertEquals("hello", context.getOne("str"));

        el = GenericArguments.seq(GenericArguments.optional(GenericArguments.integer(Translations.untr("val")), GenericArguments.string(Translations.untr("str"))));
        final CommandElement finalEl = el;
        Assertions.assertThrows(ArgumentParseException.class, () -> parseForInput("hello", finalEl));
    }

    @Test
    public void testRepeated() throws ArgumentParseException {
        CommandContext context = parseForInput("1 1 2 3 5", GenericArguments.repeated(GenericArguments.integer(Translations.untr("key")), 5));
        assertEquals(ImmutableList.<Object>of(1, 1, 2, 3, 5), context.getAll("key"));
    }

    @Test
    public void testAllOf() throws ArgumentParseException {
        CommandContext context = parseForInput("2 4 8 16 32 64 128", GenericArguments.allOf(GenericArguments.integer(Translations.untr("key"))));
        assertEquals(ImmutableList.<Object>of(2, 4, 8, 16, 32, 64, 128), context.getAll("key"));
    }

    @Test
    public void testString() throws ArgumentParseException {
        CommandContext context = parseForInput("\"here it is\"", GenericArguments.string(Translations.untr("a value")));
        assertEquals("here it is", context.getOne("a value"));
    }

    @Test
    public void testInteger() throws ArgumentParseException {
        CommandContext context = parseForInput("52", GenericArguments.integer(Translations.untr("a value")));
        assertEquals((Integer) 52, context.getOne("a value"));

        assertThrows(ArgumentParseException.class, () -> parseForInput("notanumber", GenericArguments.integer(Translations.untr("a value"))));
    }

    @Test
    public void testUUID() throws ArgumentParseException {
        final UUID subject = UUID.randomUUID();
        CommandContext ctx = parseForInput(subject.toString(), GenericArguments.uuid(Translations.untr("a value")));
        assertEquals(subject, ctx.getOne("a value"));


        assertThrows(ArgumentParseException.class, () -> parseForInput("48459", GenericArguments.uuid(Translations.untr("a value"))));
        assertThrows(ArgumentParseException.class, () -> parseForInput("words", GenericArguments.uuid(Translations.untr("a value"))));
    }

    @Test
    public void testBool() throws ArgumentParseException {
        CommandElement boolEl = GenericArguments.bool(Translations.untr("val"));
        assertEquals(true, parseForInput("true", boolEl).getOne("val"));
        assertEquals(true, parseForInput("t", boolEl).getOne("val"));
        assertEquals(false, parseForInput("f", boolEl).getOne("val"));


        assertThrows(ArgumentParseException.class, () -> parseForInput("notabool", boolEl));
    }

    private enum TestEnum {
        ONE, TWO, RED
    }

    @Test
    public void testEnumValue() throws ArgumentParseException {
        CommandElement enumEl = GenericArguments.enumValue(Translations.untr("val"), TestEnum.class);
        assertEquals(TestEnum.ONE, parseForInput("one", enumEl).getOne("val"));
        assertEquals(TestEnum.TWO, parseForInput("TwO", enumEl).getOne("val"));
        assertEquals(TestEnum.RED, parseForInput("RED", enumEl).getOne("val"));

        assertThrows(ArgumentParseException.class, () -> parseForInput("notanel", enumEl));
    }

    @Test
    public void testRemainingJoinedStrings() throws ArgumentParseException {
        CommandElement remainingJoined = GenericArguments.remainingJoinedStrings(Translations.untr("val"));
        assertEquals("one", parseForInput("one", remainingJoined).getOne("val"));
        assertEquals("one big string", parseForInput("one big string", remainingJoined).getOne("val"));
    }
}

File: core/src/test/kotlin/ca/stellardrift/permissionsex/util/command/GenericArgumentsTest.kt
/*
 * PermissionsEx
 * Copyright (C) zml and PermissionsEx contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ca.stellardrift.permissionsex.util.command

import ca.stellardrift.permissionsex.commands.parse.ArgumentParseException
import ca.stellardrift.permissionsex.commands.parse.CommandContext
import ca.stellardrift.permissionsex.commands.parse.CommandElement
import ca.stellardrift.permissionsex.commands.parse.QuotedStringParser
import ca.stellardrift.permissionsex.commands.parse.StructuralArguments
import ca.stellardrift.permissionsex.commands.parse.StructuralArguments.firstParsing
import ca.stellardrift.permissionsex.commands.parse.boolean
import ca.stellardrift.permissionsex.commands.parse.choices
import ca.stellardrift.permissionsex.commands.parse.command
import ca.stellardrift.permissionsex.commands.parse.enum
import ca.stellardrift.permissionsex.commands.parse.int
import ca.stellardrift.permissionsex.commands.parse.none
import ca.stellardrift.permissionsex.commands.parse.remainingJoinedStrings
import ca.stellardrift.permissionsex.commands.parse.string
import ca.stellardrift.permissionsex.commands.parse.uuid
import ca.stellardrift.permissionsex.util.unaryPlus
import java.util.UUID
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test

/**
 * Tests for all argument types contained in GenericArguments
 */
class GenericArgumentsTest {
    @Test
    fun testNone() {
        val args = QuotedStringParser.parseFrom("a", false)
        val context = CommandContext(NULL_COMMAND, args.raw)
        none().parse(args, context)
        assertEquals("a", args.next())
    }

    @Test
    @Disabled
    fun testFlags() {
        throw UnsupportedOperationException()
    }

    @Test
    fun testSequence() {
        val one = string() key +"one"
        val two = string() key +"two"
        val three = string() key +"three"
        val el =
            StructuralArguments.seq(one, two, three)
        val context =
            parseForInput("a b c", el)
        assertEquals("a", context[one])
        assertEquals("b", context[two])
        assertEquals("c", context[three])
        assertThrows(ArgumentParseException::class.java) { parseForInput("a b", el) }
    }

    @Test
    fun testChoices() {
        val el = choices(
            mapOf("a" to "one", "b" to "two"), +"a or b"
        ) key +"val"
        val context = parseForInput("a", el)
        assertEquals("one", context[el])
        assertThrows(ArgumentParseException::class.java) { parseForInput("c", el) }
    }

    @Test
    fun testFirstParsing() {
        val el = firstParsing(int() key +"val", string() key +"val")
        var context = parseForInput("word", el)
        assertEquals("word", context.getOne("val"))
        context = parseForInput("42", el)
        assertEquals(42, context.getOne<Int>("val"))
    }

    @Test
    fun testOptional() {
        var el = StructuralArguments.optional(string() key +"val")
        var context = parseForInput("", el)
        Assertions.assertNull(context.getOne("val"))
        // el = StructuralArguments.optional(string() key +"val", "def") // TODO: add back after re-implementation
        // context = parseForInput("", el)
        // assertEquals("def", context.getOne("val"))
        el = StructuralArguments.seq(
            StructuralArguments.optionalWeak(int() key +"val"), string() key +"str"
        )
        context = parseForInput("hello", el)
        assertEquals("hello", context.getOne("str"))
        el = StructuralArguments.seq(
            StructuralArguments.optional(
                int() key +"val"), string() key +"str"
            )
        assertThrows(ArgumentParseException::class.java) {
            parseForInput("hello", el)
        }
    }

    @Test
    @Throws(ArgumentParseException::class)
    fun testRepeated() {
        val context =
            parseForInput(
                "1 1 2 3 5",
                StructuralArguments.repeated(int() key +"key", 5)
            )
        assertEquals(
            listOf(1, 1, 2, 3, 5),
            context.getAll<Any>("key")
        )
    }

    @Test
    @Throws(ArgumentParseException::class)
    fun testAllOf() {
        val context =
            parseForInput(
                "2 4 8 16 32 64 128",
                StructuralArguments.allOf(int() key +"key")
            )
        assertEquals(
            listOf(2, 4, 8, 16, 32, 64, 128),
            context.getAll<Any>("key")
        )
    }

    @Test
    @Throws(ArgumentParseException::class)
    fun testString() {
        val context =
            parseForInput(
                "\"here it is\"",
                string() key +"a value"
            )
        assertEquals("here it is", context.getOne("a value"))
    }

    @Test
    @Throws(ArgumentParseException::class)
    fun testInteger() {
        val context = parseForInput("52", int() key +"a value")
        assertEquals(52, context.getOne<Int>("a value"))
        assertThrows(
            ArgumentParseException::class.java
        ) {
            parseForInput(
                "notanumber",
                int() key +"a value"
            )
        }
    }

    @Test
    @Throws(ArgumentParseException::class)
    fun testUUID() {
        val subject = UUID.randomUUID()
        val ctx =
            parseForInput(
                subject.toString(),
                uuid() key +"a value"
            )
        assertEquals(subject, ctx.getOne("a value"))
        assertThrows(
            ArgumentParseException::class.java
        ) {
            parseForInput(
                "48459",
                uuid() key +"a value"
            )
        }
        assertThrows(
            ArgumentParseException::class.java
        ) {
            parseForInput(
                "words",
                uuid() key +"a value"
            )
        }
    }

    @Test
    @Throws(ArgumentParseException::class)
    fun testBool() {
        val boolEl = boolean() key +"val"
        assertEquals(true, parseForInput("true", boolEl)[boolEl])
        assertEquals(true, parseForInput("t", boolEl)[boolEl])
        assertEquals(false, parseForInput("f", boolEl)[boolEl])
        assertThrows(ArgumentParseException::class.java) {
            parseForInput("notabool", boolEl)
        }
    }

    private enum class TestEnum {
        ONE, TWO, RED
    }

    @Test
    @Throws(ArgumentParseException::class)
    fun testEnumValue() {
        val enumEl = enum<TestEnum>() key +"val"
        assertEquals(
            TestEnum.ONE,
            parseForInput("one", enumEl)[enumEl]
        )
        assertEquals(
            TestEnum.TWO,
            parseForInput("TwO", enumEl)[enumEl]
        )
        assertEquals(
            TestEnum.RED,
            parseForInput("RED", enumEl)[enumEl]
        )
        assertThrows(ArgumentParseException::class.java) {
            parseForInput("notanel", enumEl)
        }
    }

    @Test
    @Throws(ArgumentParseException::class)
    fun testRemainingJoinedStrings() {
        val remainingJoined = remainingJoinedStrings() key +"val"
        assertEquals("one",
            parseForInput("one", remainingJoined)[remainingJoined]
        )
        assertEquals("one big string",
            parseForInput("one big string", remainingJoined)[remainingJoined]
        )
    }
}

private val NULL_COMMAND =
    command("test") { executor { _, _ -> } }

@Throws(ArgumentParseException::class)
private fun parseForInput(
    input: String,
    element: CommandElement
): CommandContext {
    val args = QuotedStringParser.parseFrom(input, false)
    val context = CommandContext(NULL_COMMAND, args.raw)
    element.parse(args, context)
    return context
}
